### ISSUE TITLE
Further memory leak (probably in C code)

### ISSUE THREAD

**Issue by DavAug** [MichaelClerx/myokit#859]
Following the fix #855 I still experienced an explosion of memory usage in my inference algorithm which still seems to be related to `myokit.Simulation.run()`. In contrast to the protocol, I can, however, no longer see the growing memory in Python objects, i.e. using pympler as in #855 . But using the same example

```
import myokit

m, p, x = myokit.load('example')
s = myokit.Simulation(m, p)

for _ in range(100000):
    s.run(1000)
```

and the command line process profiler 

```[bash]
top
```

I can see that over time the memory usage grows / the available memory decreases linearly in time, i.e. proportional to the number of `myokit.Simulation.run()` calls. Here are two snapshots 2 minutes apart and we can see that the available memory reduces from 14672.1 to 14640.7 / that the memory usage of the python process increases from 0.3% of the total memory to 0.5% of the total memory.

```
top - 10:10:47 up 2 days,  3:43,  1 user,  load average: 0.28, 0.15, 0.24
Tasks: 244 total,   2 running, 242 sleeping,   0 stopped,   0 zombie
%Cpu(s): 12.5 us,  0.0 sy,  0.0 ni, 87.4 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
MiB Mem :  15772.3 total,  12875.2 free,    688.6 used,   2208.4 buff/cache
MiB Swap:  32768.0 total,  32768.0 free,      0.0 used.  14672.1 avail Mem

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
  15964 davtin    20   0  569132  46200  15344 R  99.7   0.3   0:17.14 python
```

```
top - 10:12:47 up 2 days,  3:45,  1 user,  load average: 0.91, 0.44, 0.34
Tasks: 244 total,   2 running, 242 sleeping,   0 stopped,   0 zombie
%Cpu(s): 12.5 us,  0.0 sy,  0.0 ni, 87.5 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
MiB Mem :  15772.3 total,  12843.8 free,    720.0 used,   2208.5 buff/cache
MiB Swap:  32768.0 total,  32768.0 free,      0.0 used.  14640.7 avail Mem

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
  15964 davtin    20   0  597124  74184  15344 R  99.7   0.5   2:17.18 python
```

For my (bigger) model this is enough to kill the processes after 12h because the computer runs out of memory. Do you have an idea how we might tackle this?

**Comment by MichaelClerx** [comment]
I'd start by forcing the garbage collector and seeing what that changes

```
import gc
for i in range(100000):
    s.run(1000)
    if i % 10 == 0:
        gc.collect()
```

**Comment by DavAug** [comment]
> I'd start by forcing the garbage collector and seeing what that changes
> 
> ```
> import gc
> for i in range(100000):
>     s.run(1000)
>     if i % 10 == 0:
>         gc.collect()
> ```

That does not seem to change the memory consumption. I'll try to have a closer look at the memory allocations :)

**Comment by MichaelClerx** [comment]
Thanks @DavAug !

I've been doing some googling on clever ways to fix it (because I always go the `dumb' re-read my code and look for errors route). It seems using `valgrind` is the way to go, although I've never tried this myself. I think Martin & Fergus might be good people to ask about this? One thing I noticed is that you really need to check the publication date of the info you find: older blog posts etc. want  you to recompile python with some special flags (don't!), while since Python 3.4 it seems all you need to do is set an environment variable (and download some file that tells valgrind to suppress errors that are in Python itself, not in the extension module).

This also looks promising: https://pypi.org/project/pymemtrace/

And you could also think of going the simple route, and e.g. carefully read through the code, or comment bits out and see if that helps etc. to manually pinpoint the error (not necessarily as stupid as it might sound, because this kind of low-tech solution removes instead of adds complexity)

**Comment by MichaelClerx** [comment]
It might actually be very hard to use clever tools, as the C code and library don't exist yet when you call your script :D 
(Unless we fix #294  first)

**Comment by DavAug** [comment]
> It might actually be very hard to use clever tools, as the C code and library don't exist yet when you call your script :D (Unless we fix #294 first)

Yes I figured that that might be difficult. I am in the process of looking carefully at the `.sim_clean()` method in cvodessim.c to see whether some simulation related variables are not fully removed. 

#294 sounds very cool but atm I wouldn't really know how to do it :D

**Comment by DavAug** [comment]
I finally found some time again to come back to this and I don't think I fully understand what is happening :D

As far as I can see there are a number of globally defined variables:
```
int initialised = 0; /* Has the simulation been initialised */
Model model;        /* A model object */
ESys epacing;           /* Event-based pacing system */
PyObject* eprotocol;    /* An event-based pacing protocol */
FSys fpacing;           /* Fixed-form pacing system */
PyObject* fprotocol;    /* A fixed-form pacing protocol */
double pace = 0;        /* Pacing value */
void *cvode_mem;     /* The memory used by the solver */
#if SUNDIALS_VERSION_MAJOR >= 3
SUNMatrix sundense_matrix;          /* Dense matrix for linear solves */
SUNLinearSolver sundense_solver;    /* Linear solver object */
#endif
#if SUNDIALS_VERSION_MAJOR >= 6
SUNContext sundials_context; /* A sundials context to run in (for profiling etc.) */
#endif
UserData udata;      /* UserData struct, used to pass in parameters */
realtype* pbar;      /* Vector of independents in user data */
double abs_tol = 1e-6;  /* The absolute tolerance */
double rel_tol = 1e-4;  /* The relative tolerance */
double dt_max = 0;      /* The maximum step size (0.0 for none) */
double dt_min = 0;      /* The minimum step size (0.0 for none) */
double realtime = 0;        /* Time since start */
long evaluations = 0;       /* Number of evaluations since sim init */
long steps = 0;             /* Number of steps since sim init */
int zero_step_count;
const int max_zero_step_count = 500;
N_Vector y;     /* The current position y */
N_Vector* sy;   /* Current state sensitivities, 1 vector per independent */
N_Vector z;
N_Vector* sz;
N_Vector ylast;
PyObject* literals;     /* A list of literal constant values */
PyObject* parameters;   /* A list of parameter values */
PyObject* state_py;     /* List: The state passed from and to Python */
PyObject* s_state_py;   /* List: The state sensitivities passed from and to Python */
PyObject* bound_py;     /* List: The bound variables, passed to Python */
double t;       /* Current simulation time */
double tlast;   /* Previous simulation time, for error and progress tracking */
double tnext;   /* Next simulation halting point */
double tmin;    /* The initial simulation time */
double tmax;    /* The final simulation time */
int dynamic_logging;    /* True if logging every point. */
PyObject* log_dict;     /* The log dict (DataLog) */
PyObject* sens_list;    /* Sensitivity logging list */
double tlog;            /* Next time to log */
double log_interval;    /* The periodic logging interval */
Py_ssize_t ilog;        /* Index of next point in the point list */
PyObject* log_times;    /* The point list (or None if disabled) */
int rf_indice;          /* Indice of state variable to use in root finding (ignored if not enabled) */
double rf_threshold;    /* Threshold to use for root finding (ignored if not enabled) */
PyObject* rf_list;      /* List to store found roots in (or None if not enabled) */
int* rf_direction;      /* Direction of root crossings: 1 for up, -1 for down, 0 for no crossing. */
PyObject* benchmarker;      /* myokit.tools.Benchmarker object */
PyObject* benchmarker_time_str;
int log_realtime;           /* 1 iff we're logging real simulation time */
double realtime_start;      /* time when sim run started */
```

There are also various local variables in the ``.sim_init()`` and ``.sim_clean()`` methods, but let's assume that those don't give rise to permanent memory allocations.

Some of the global variables are assigned with ``NULL`` in the beginning of ``.sim_init()`` which probably frees the allocated memory (?)
```
    model = NULL;
    epacing = NULL;
    fpacing = NULL;
    udata = NULL;
    pbar = NULL;
    y = NULL;
    sy = NULL;
    z = NULL;
    sz = NULL;
    ylast = NULL;
    log_times = NULL;
    benchmarker_time_str = NULL;
    #ifdef MYOKIT_DEBUG_PROFILING
    benchmarker_print_str = NULL;
    #endif
    cvode_mem = NULL;
    #if SUNDIALS_VERSION_MAJOR >= 3
    sundense_matrix = NULL;
    sundense_solver = NULL;
    #endif
    #if SUNDIALS_VERSION_MAJOR >= 6
    sundials_context = NULL;
    #endif
```

After this the CVODES does it's magic and there are two ways how the calls can be returned: 1. ``.sim_clean()`` and 2. ``.sim_cleanx()``. The latter is only used if some compilation initialisation error is thrown, so that can't really be the problem here, because the model is evaluated many times without errors. The ``.sim_clean()`` method takes care of the following variables:

```
y
sy
z
sz
rf_direction
cvode_mem
sundense_solver
sundense_matrix
sundials_context
pbar
udata
epacing
fpacing
model
benchmarker_print_str
```

Assuming that the ``.sim_clean()`` method free all the variables above without problems, then this leaves the following global variables that are not freed after a simulation run

```
initialised
eprotocol
fprotocol
pace
abs_tol
rel_tol
dt_max
dt_min
realtime
evaluations
long steps
zero_step_count
max_zero_step_count
ylast
literals
parameters
state_py
s_state_py
bound_py
t
tlast
tnext
tmin
tmax
dynamic_logging
log_dict
sens_list
tlog
log_interval
ilog
log_times
rf_indice
rf_threshold
rf_list
rf_direction
benchmarker
log_realtime
realtime_start
```

I can definitely see that a bunch of these variables don't need to be freed because they remain the same for all simulation runs or are overwritten by equally sized inputs. But is it possible that there are some variables that keep growing in size? For example I am not sure that ``evaluations`` is nulled after execution. So maybe ``evaluations`` then just keeps growing.

Let me know what you think @MichaelClerx . I am a little clueless :D
